github_backup/github_backup.py

```diff
@@ -0,0 +1,1143 @@
+            name,
+            masked_remote_url,
+            local_dir))
+        if bare_clone:
+            if lfs_clone:
+                git_command = ['git', 'lfs', 'clone', '--mirror', remote_url, local_dir]
+            else:
+                git_command = ['git', 'clone', '--mirror', remote_url, local_dir]
+        else:
+            if lfs_clone:
+                git_command = ['git', 'lfs', 'clone', remote_url, local_dir]
+            else:
+                git_command = ['git', 'clone', remote_url, local_dir]
+        logging_subprocess(git_command, None)
+
+
+def backup_account(args, output_directory):
+    account_cwd = os.path.join(output_directory, 'account')
+
+    if args.include_starred or args.include_everything:
+        output_file = "{0}/starred.json".format(account_cwd)
+        template = "https://{0}/users/{1}/starred".format(get_github_api_host(args), args.user)
+        _backup_data(args,
+                     "starred repositories",
+                     template,
+                     output_file,
+                     account_cwd)
+
+    if args.include_watched or args.include_everything:
+        output_file = "{0}/watched.json".format(account_cwd)
+        template = "https://{0}/users/{1}/subscriptions".format(get_github_api_host(args), args.user)
+        _backup_data(args,
+                     "watched repositories",
+                     template,
+                     output_file,
+                     account_cwd)
+
+    if args.include_followers or args.include_everything:
+        output_file = "{0}/followers.json".format(account_cwd)
+        template = "https://{0}/users/{1}/followers".format(get_github_api_host(args), args.user)
+        _backup_data(args,
+                     "followers",
+                     template,
+                     output_file,
+                     account_cwd)
+
+    if args.include_following or args.include_everything:
+        output_file = "{0}/following.json".format(account_cwd)
+        template = "https://{0}/users/{1}/following".format(get_github_api_host(args), args.user)
+        _backup_data(args,
+                     "following",
+                     template,
+                     output_file,
+                     account_cwd)
+
+
+def _backup_data(args, name, template, output_file, output_directory):
+    skip_existing = args.skip_existing
+    if not skip_existing or not os.path.exists(output_file):
+        log_info('Retrieving {0} {1}'.format(args.user, name))
+        mkdir_p(output_directory)
+        data = retrieve_data(args, template)
+
+        log_info('Writing {0} {1} to disk'.format(len(data), name))
+        with codecs.open(output_file, 'w', encoding='utf-8') as f:
+            json_dump(data, f)
+
+
+def json_dump(data, output_file):
+    json.dump(data,
+              output_file,
+              ensure_ascii=False,
+              sort_keys=True,
+              indent=4,
+              separators=(',', ': '))
+
+
+def main():
```

Comment:
This function should be wholly moved to the `bin`.

Reply:
Is this from a separation of concerns standpoint? I don't remember this coming up in my last attempt at this, but will try my best. I want to make sure I understand what's being asked, and how to achieve both of our goals. I assume your goal is to not break compatibility or functionality. Mine is to use this package inside another one.

My understanding is that the purpose of `main()` is to digest arguments, setup the environment, and execute some functions to achieve an end goal.

With the understanding that `main()` will be defined inside `bin/github-backup`, which of these would you prefer, if either meets your goals?
* Have `main()` be a wrapper to another function, say `execute_backup()` located in `github_backup/github_backup.py`, that digests arguments, sets up the environment, and executes functions to achieve the end goal
* Have `main()` continue to do all of that itself, and others (e.g. me) that are importing the library implement their own flavor of "main()", since it won't be (easily) importable?

Reply:
I reread your comment, and went with the second option I proposed. Let me know if this works :)